source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc

```diff
@@ -116,12 +126,22 @@ void GrpcClientImpl::toAuthzResponseHeader(
     ResponsePtr& response,
     const Protobuf::RepeatedPtrField<envoy::config::core::v3::HeaderValueOption>& headers) {
   for (const auto& header : headers) {
-    if (header.append().value()) {
```

Comment:
From my understanding, `header.append().value()` will be true if and only if the append field was was present and had a true value. In this case, the old code would add to headers_to_append (i.e APPEND_IF_EXISTS_OR_ADD). In all other cases (e.g. field is missing or field is explicitly set to false) it would add to headers_to_set (i.e, OVERWRITE_IF_EXISTS_OR_ADD).

The new code gets an append_action from Http::HeaderUtility::getHeaderAppendAction(header). If the append field is not present, then the default action is returned (APPEND_IF_EXISTS_OR_ADD). I think this doesn't match the existing behavior, so I could either be missing something about how the new code works of my understanding is off - what do you think?

Reply:
Hmmm. Current behavior of when "append" is missing seems to be "true" everywhere else (Mutations etc.) which means that "APPEND_IF_EXISTS_OR_ADD" should be the right action given we want to keep things consistent.
That said, I did come across a comment assuming opposite of this and forcing "OVERWRITE" when "append" is missing (perhaps a bug?)

@esmet I'm okay either ways in case you want to change it back to "OVERWRITE" instead of "APPEND" when the "append" field is not specified.

![image](https://user-images.githubusercontent.com/12120828/135651888-1dfa1218-02c1-4e48-8067-4228d7c5f49f.jpeg)

Reply:
The documentation definitely does say that the default is true, but I don't think that's totally true in practice. I see three places `append` is used:

ext_authz/ext_authz_grpc_impl.cc (here) Uses the default value from protobuf (false)
```
          if (header.append().value()) {
              ...
```

ext_proc/mutation_utils.cc defaults to false with a comment noting ext_authz as precedent
```
      // Make "false" the default. This is logical and matches the ext_authz
      // filter. However, the router handles this same protobuf and uses "true"
      // as the default instead.
      const bool append = PROTOBUF_GET_WRAPPED_OR_DEFAULT(sh, append, false);
```

router/header_parser.cc defaults to true
```
const bool append = PROTOBUF_GET_WRAPPED_OR_DEFAULT(header_value_option, append, true);
```

So, I wonder if the comment was originally authored when the router was the only meaningful user of the 'append' field on HeaderValueOption. Later, when ext_authz and ext_proc adopted it, they chose the more obvious default but the comment then became stale.

If all of this makes sense (and is correct) then I think we need to update the comment for `append` to explain what's going on. Then, we need to update the ext_authz code paths to make sure there isn't a behavioral change.

Reply:
Make sense. So in the context of this PR, should we default to "OVERWRITE_IF_EXISTS_OR_ADD" here instead of "APPEND_IF_EXISTS_OR_ADD"?

Reply:
I believe the default needs to remain OVERWRITE_IF_EXISTS_OR_ADD. Ideally, we would have a unit test that verifies this behavior when omitting `append` before any of the changes in this PR. Then, when this PR adds a new field, the test should still pass as it does not set `append` nor `append_action`.